File: roboplan_examples/python/teleop/command_filter.py
import numpy as np

from .config import TeleopConfig
from .types import TwistCommand


class CommandFilter:
    """Applies deadzone, scaling, clipping, and smoothing to teleop commands."""

    def __init__(self):
        self._prev_linear = np.zeros(3, dtype=float)
        self._prev_angular = np.zeros(3, dtype=float)

    def reset(self) -> None:
        self._prev_linear[:] = 0.0
        self._prev_angular[:] = 0.0

    def _apply_deadzone(self, values: np.ndarray, threshold: float) -> np.ndarray:
        out = values.copy()
        out[np.abs(out) < threshold] = 0.0
        return out

    def filter(self, raw_cmd: TwistCommand, config: TeleopConfig) -> TwistCommand:
        linear = self._apply_deadzone(
            np.asarray(raw_cmd.linear_xyz, dtype=float), config.linear_deadzone
        )
        angular = self._apply_deadzone(
            np.asarray(raw_cmd.angular_xyz, dtype=float), config.angular_deadzone
        )

        linear = np.clip(linear, -1.0, 1.0) * config.linear_sensitivity
        angular = np.clip(angular, -1.0, 1.0) * config.angular_sensitivity

        linear = np.clip(linear, -config.max_linear_speed, config.max_linear_speed)
        angular = np.clip(angular, -config.max_angular_speed, config.max_angular_speed)

        alpha = float(np.clip(config.smoothing_alpha, 0.0, 1.0))
        linear = alpha * linear + (1.0 - alpha) * self._prev_linear
        angular = alpha * angular + (1.0 - alpha) * self._prev_angular

        self._prev_linear = linear
        self._prev_angular = angular

        return TwistCommand(
            linear_xyz=linear,
            angular_xyz=angular,
            gripper=float(np.clip(raw_cmd.gripper, -1.0, 1.0)),
            frame=raw_cmd.frame,
            stamp_ns=raw_cmd.stamp_ns,
        )

File: roboplan_oink/src/teleop/command_filter.cpp
#include <roboplan_oink/teleop/command_filter.hpp>

#include <algorithm>
#include <cmath>

namespace roboplan {

CommandFilter::CommandFilter(const CommandFilterConfig& config) : config_(config) {}

void CommandFilter::applyDeadzone(Eigen::Ref<Eigen::Vector3d> values, double threshold) {
  for (int i = 0; i < values.size(); ++i) {
    if (std::abs(values[i]) < threshold) {
      values[i] = 0.0;
    }
  }
}

Eigen::Matrix<double, 6, 1>
CommandFilter::filter(const Eigen::Ref<const Eigen::Matrix<double, 6, 1>>& raw_twist) {
  Eigen::Vector3d linear = raw_twist.head<3>();
  Eigen::Vector3d angular = raw_twist.tail<3>();

  applyDeadzone(linear, config_.linear_deadzone);
  applyDeadzone(angular, config_.angular_deadzone);

  linear = linear.cwiseMax(-1.0).cwiseMin(1.0) * config_.linear_sensitivity;
  angular = angular.cwiseMax(-1.0).cwiseMin(1.0) * config_.angular_sensitivity;

  linear = linear.cwiseMax(-config_.max_linear_speed).cwiseMin(config_.max_linear_speed);
  angular = angular.cwiseMax(-config_.max_angular_speed).cwiseMin(config_.max_angular_speed);

  const double alpha = std::clamp(config_.smoothing_alpha, 0.0, 1.0);
  linear = alpha * linear + (1.0 - alpha) * prev_linear_;
  angular = alpha * angular + (1.0 - alpha) * prev_angular_;

  prev_linear_ = linear;
  prev_angular_ = angular;

  Eigen::Matrix<double, 6, 1> filtered;
  filtered << linear, angular;
  return filtered;
}

void CommandFilter::reset() {
  prev_linear_.setZero();
  prev_angular_.setZero();
}

void CommandFilter::setConfig(const CommandFilterConfig& config) { config_ = config; }

const CommandFilterConfig& CommandFilter::getConfig() const { return config_; }

}  // namespace roboplan
